### PR TITLE
Use 64-bit file I/O functions, for Android compatibility

### PR DESCRIPTION
**Pull request by qmfrederik**
On 32-bit Android, the file functions like lseek also operate on 32-bit values. `_FILE_OFFSET_BITS` and friends have no impact on Android, see the 32-bit notes at the end of the [bionic README.md](https://android.googlesource.com/platform/bionic.git/+/master/README.md)

This PR updates the codebase to always use 64-bit file I/O functions like `lseek64`. This effectively reverts #736.

### CODE REVIEW

**Comment by qmfrederik** [comment]
/cc @stephentoub as this reverts #736. 

I did not rename the PAL functions, let me know if you want me to change that too.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Ubuntu14.04 Debug please
@dotnet-bot test Outerloop OSX Debug please
@dotnet-bot test outerloop PortableLinux Debug
@dotnet-bot test outerloop OpenSUSE42.1 Release
@dotnet-bot test outerloop CentOS7.1 Debug
@dotnet-bot test outerloop RHEL7.2 Release

**Comment by stephentoub** [comment]
> I did not rename the PAL functions

Leaving them as-is is fine.

> as this reverts #736

Yeah, that was from before we had the shims.

**Comment by qmfrederik** [comment]
ok - so it looks like at least macOS doesn't have the fseek64 functions and friends.

I can think of a couple of ways around it:

- CMake checks and HAVE_FSEEK64 preprocessor variables, and then #if HAVE_FSEEK64/#else blocks
- In pal_io.cpp, something like: 
```
#ifndef fseek64
#define fseek64 fseek
#endif
```
- A mix of both

I'll let the builds finish first to see if it detects other errors.

Before I update the PR, any preferences?

**Comment by stephentoub** [comment]
> I can think of a couple of ways around it

I'd personally prefer the cmake / HAVE_FSEEK64 if/else approach, as that's how we've done it elsewhere in the repo.

**Comment by qmfrederik** [comment]
@stephentoub Is this what you had in mind?

File: src/Native/Unix/System.Native/pal_io.cpp [hunk 1]
```diff
@@ -1140,9 +1179,9 @@ extern "C" int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destination
         // futimes is not a POSIX function, and not available on Android,
         // but futimens is
         struct timespec origTimes[2];
-        origTimes[0].tv_sec = sourceStat.st_atime;
+        origTimes[0].tv_sec = static_cast<__kernel_time_t>(sourceStat.st_atime);
         origTimes[0].tv_nsec = 0;
-        origTimes[1].tv_sec = sourceStat.st_mtime;
+        origTimes[1].tv_sec = static_cast<__kernel_time_t>(sourceStat.st_mtime);
```
Comment (stephentoub) [inline src/Native/Unix/System.Native/pal_io.cpp]: Why __kernel_time_t?

Comment (qmfrederik) [inline src/Native/Unix/System.Native/pal_io.cpp]: You're right, it should be time_t; __kernel_time_t is what came up in the compiler warnings - fixed it.

**Comment by qmfrederik** [comment]
@stephentoub Should be all good now. Thanks for the review!

**Pull request merged**